### PR TITLE
#3165 cadproject file reviews - Create test data for parts

### LINKED ISSUE #3165: [CAD/Project File Reviews] - Create Test Data for Parts
### Description

We want to have generic seed data that can be used for testing future endpoints/frontend features. This should consist of utils functions to create a part as well as a variety of different examples of parts using this function. Creating submissions and reviews are in a different ticket, so those fields can be left as an empty array for the time being.

### Acceptance Criteria

There should be a utils function to create parts in the test-utils.ts file. There should then be examples made in the seed data. There should be a range of different examples including with and without optional fields, and other applicable edge cases. The submissions field can be left as an empty array for this ticket. Examples should have informative names.

### Proposed Solution

Create functions in test-utils.ts file that take in all fields for each model, and create prisma objects using that data. Additionally, create example parts in a parts.seed.ts file in prisma/seed-data, which are of type Prisma.PartCreateArgs. These should then be used in seed.ts to create all the examples

### Mocks

_No response_

### PR DESCRIPTION
**Pull request by mokaung**
## Changes

- Created test function createTestPart in test-utils.ts
- Created test objects in parts.seed.ts
- Created examples in seed.ts

## Notes

Please let me know if there are things I need to fix, or if I did something the wrong way!
Also let me know if I should write more tests
I wasn't able to run yarn test:backend since for some reason it said the command didn't exist.

## Test Cases

- Basic Part
- Part without desription
- Part without Image
- Part with empty history
- Part with long name
- Part with negative index
- Part with index zero
- Part with large index
- Part with 'Ready to Review' tag
- Part with 'In Review' tag
- Part with 'Reviewed' tag
- Part with 'Approved' tag
- Part with current date
- Part with past date
- Part with Unix Epoch date
- Part with future date
- Part with Leap year date

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ x ] All commits are tagged with the ticket number
- [ x ] No linting errors / newline at end of file warnings
- [ x ] All code follows repository-configured prettier formatting
- [ x ] No merge conflicts
- [ x ] All checks passing
- [ x ] Screenshots of UI changes (see Screenshots section)
- [ x ] Remove any non-applicable sections of this template
- [ x ] Assign the PR to yourself
- [ x ] No `yarn.lock` changes (unless dependencies have changed)
- [ x ] Request reviewers & ping on Slack
- [ x ] PR is linked to the ticket (fill in the closes line below)

Closes #3165

### CODE REVIEW

File: src/backend/src/prisma/seed-data/parts.seed.ts [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+  return {
+    data: {
+      index: 1,
+      commonName: 'Basic Part',
+      description: 'Basic part with all fields populated',
+      previewImageLink: 'https://NER.com/basicpart.jpg',
+      status: 'IN_PROGRESS',
+      createdAt: new Date('2025-01-01T10:00:00Z'),
+      history: ['Created part', 'Updated specs', 'Assigned to review', 'Reviewed part'],
+      projectId,
```
Comment (chpy04) [inline src/backend/src/prisma/seed-data/parts.seed.ts]: use a connect statement here to make it clear that this is a relation

File: src/backend/src/prisma/seed-data/parts.seed.ts [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+      userCreatedId,
```
Comment (chpy04) [inline src/backend/src/prisma/seed-data/parts.seed.ts]: same thing here

File: src/backend/src/prisma/seed-data/parts.seed.ts [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+      assignees: {
+        connect: [{ userId: 'REPLACE_WITH_USER_ID' }] // to be replaced in seed.ts
```
Comment (chpy04) [inline src/backend/src/prisma/seed-data/parts.seed.ts]: Similar to how you take in a projected and userCreatedId, take in user Ids for assignees and reviewers and connect them here

File: src/backend/src/prisma/seed-data/parts.seed.ts [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+      },
+      reviewers: {
+        connect: [{ userId: 'REPLACE_WITH_USER_ID' }] // to be replaced in seed.ts
```
Comment (chpy04) [inline src/backend/src/prisma/seed-data/parts.seed.ts]: Same thing as above

File: src/backend/src/prisma/seed-data/parts.seed.ts [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+      }
+    }
+  };
+};
```
Comment (chpy04) [inline src/backend/src/prisma/seed-data/parts.seed.ts:364]: Make sure to make the changes above for all of these examples

File: src/backend/src/prisma/seed.ts [hunk 6]
```diff
@@ -2131,6 +2132,20 @@ const performSeed: () => Promise<void> = async () => {
...
+  /**
+   * PARTS
+   */
+  for (const createPart of Object.values(dbSeedAllParts)) {
+    const partArgs = createPart(project2Id, batman.userId);
+    await prisma.part.create({
+      data: {
+        ...partArgs.data,
+        assignees: { connect: [{ userId: hawkMan.userId }] },
+        reviewers: { connect: [{ userId: cyborg.userId }] }
```
Comment (chpy04) [inline src/backend/src/prisma/seed.ts]: Can you have thomas ermax be the reviewer on some, and requester on some other ones? This will make it easier when building out the frontend because people won't have to be on different user accounts to see both functionality

File: src/backend/tests/test-utils.ts [hunk 7]
```diff
@@ -604,6 +605,37 @@ export const createTestTask = async (
...
+      submissions: { create: [] },
+      project: { connect: { projectId } },
+      assignees: { connect: assignees.map((user) => ({ userId: user.userId })) },
+      reviewers: { connect: reviewers.map((user) => ({ userId: user.userId })) },
+      history,
+      createdAt,
+      userCreated: { connect: { userId: userCreatedId } }
+    }
+  });
+};
```
Comment (chpy04) [inline src/backend/tests/test-utils.ts]: Due to some changes about how we are going to go about the order of development, we actually don't need this function anymore, so can you get rid of the changes on this file? Sorry for the confusion this is completely on me.

**Review by chpy04 — CHANGES_REQUESTED** [review]

**Review by walker-sean — CHANGES_REQUESTED** [review]
Looks like you accidentally committed a `.zip` - also, don't forget to fix the merge conflicts plz

File: src/backend/src/prisma/seed-data/parts.seed.ts (hunk 1 above)
Comment (mokaung) [reply src/backend/src/prisma/seed-data/parts.seed.ts]: implemented changes!

File: src/backend/src/prisma/seed-data/parts.seed.ts (hunk 2 above)
Comment (mokaung) [reply src/backend/src/prisma/seed-data/parts.seed.ts]: implemented changes!

File: src/backend/src/prisma/seed-data/parts.seed.ts (hunk 3 above)
Comment (mokaung) [reply src/backend/src/prisma/seed-data/parts.seed.ts]: implemented changes!

File: src/backend/src/prisma/seed-data/parts.seed.ts (hunk 4 above)
Comment (mokaung) [reply src/backend/src/prisma/seed-data/parts.seed.ts]: implemented changes!

File: src/backend/src/prisma/seed-data/parts.seed.ts (hunk 5 above)
Comment (mokaung) [reply src/backend/src/prisma/seed-data/parts.seed.ts:364]: implemented changes!

File: src/backend/src/prisma/seed.ts (hunk 6 above)
Comment (mokaung) [reply src/backend/src/prisma/seed.ts]: implemented changes!

File: src/backend/tests/test-utils.ts (hunk 7 above)
Comment (mokaung) [reply src/backend/tests/test-utils.ts]: implemented changes!

**Comment by mokaung** [comment]
removed .zip, and fixed requested changes. Please let me know further if there are more issues, or if I implemented the fixes incorrectly

File: src/backend/src/prisma/seed.ts [hunk 8]
```diff
@@ -49,6 +49,8 @@ import OrganizationsService from '../services/organizations.services';
 import { seedGraph } from './seed-data/statistics.seed';
 import AnnouncementService from '../services/announcement.service';
 import OnboardingServices from '../services/onboarding.services';
+import { dbSeedAllParts } from './seed-data/parts.seed';
+
 import { dbSeedAllPartTags } from './seed-data/parts.seed';
```
Comment (chpy04) [inline src/backend/src/prisma/seed.ts]: combine these imports

File: src/backend/src/prisma/seed.ts [hunk 9]
```diff
@@ -2142,6 +2144,16 @@ const performSeed: () => Promise<void> = async () => {
     false
   );
 
+  /**
+   * PARTS
+   */
+  let i = 0;
+  for (const createPart of Object.values(dbSeedAllParts)) {
```
Comment (chpy04) [inline src/backend/src/prisma/seed.ts]: Im a little confused, where are the parts with the createPart args made? Ideally we should be able to just pass these into a prisma.part.Create, and then we wouldn't need to use the test functions in the seed file. If I'm missing something let me know

File: src/backend/src/prisma/seed-data/parts.seed.ts [hunk 10]
```diff
@@ -1,9 +1,386 @@
-/*
- * This file is part of NER's FinishLine and licensed under GNU AGPLv3.
+/* This file is part of NER's FinishLine and licensed under GNU AGPLv3.
```
Comment (chpy04) [inline src/backend/src/prisma/seed-data/parts.seed.ts]: get rid of this change

File: src/backend/src/prisma/migrations/migration_lock.toml [hunk 11]
```diff
@@ -1,3 +1,3 @@
 # Please do not edit this file manually
-# It should be added in your version-control system (i.e. Git)
+# It should be added in your version-control system (e.g., Git)
```
Comment (chpy04) [inline src/backend/src/prisma/migrations/migration_lock.toml]: get rid of this change

**Review by chpy04 — CHANGES_REQUESTED** [review]
few minor changes

File: src/backend/src/prisma/seed.ts (hunk 9 above)
Comment (mokaung) [reply src/backend/src/prisma/seed.ts]: resolved in meeting!

Comment (mokaung) [reply src/backend/src/prisma/seed.ts]: oh yeah i changed the named from 'createPart' to 'testPart'. I was trying to initiate previous .seed.ts files in the way they create test objects, so every 'createPart' was a function, which uses Prisma.PartCreateArgs, not the now removed createTestPart function which existed in test-utils.ts file.

File: src/backend/src/prisma/migrations/migration_lock.toml (hunk 11 above)
Comment (mokaung) [reply src/backend/src/prisma/migrations/migration_lock.toml]: resolved

File: src/backend/src/prisma/seed-data/parts.seed.ts (hunk 10 above)
Comment (mokaung) [reply src/backend/src/prisma/seed-data/parts.seed.ts]: resolved

**Review by chpy04 — APPROVED** [review]
LGTM, @walker-sean ?

**Review by walker-sean — APPROVED** [review]
